### PR TITLE
Fix possible Max audio buffering reached bug if ASIO driver takes too long to respond

### PR DESCRIPTION
**Pull request by marcelofg55**
This fixes a possible "Max audio buffering reached" problem when an ASIO driver takes too long to respond, which can lead to the audio track being silent since that point.

I could reproduce this bug with a Focusrite Scarlett 6i6 1st gen when starting REAPER, here's a small video showing the issue: https://gfycat.com/constantcavernousduck

### CODE REVIEW

**Comment by Andersama** [comment]
Can't exactly make out what's going on in that clip you posted. The max audio buffering is an extremely old obs (annoying) behavior has where it would disable audio devices if they "buffered" for too long. It makes sense to some degree if the device is otherwise just not available and to not waste cycles on it, but it may be exactly that, the device may not have been responsive.

If you're suggesting you do / don't get the max audio buffering issue based on this change in code let me know. The audio being silent after max buffering is exactly what obs does, that's exactly what I'd expect. If this code seems prevent it from hitting the max buffer let me know.

**Comment by marcelofg55** [comment]
Hi Andersama, yes, that's exactly what this code does, after these changes I couldn't reproduce the silent track bug anymore. As I said in the first post before I could reproduce the bug everytime I started REAPER, which lead to a silent track in the recorded video. After these changes the "Max audio buffering reached" doesn't occur when starting REAPER anymore, so it fixes the silent track issue.

Btw, the GIF clip I posted is without these changes, I posted it just to show the bug as it happened before.

**Comment by Andersama** [comment]
Just to be clear the way this is working is you're allowing the audio thread to write data from previous frames in a circle buffer in the event the device doesn't respond in time. in theory if a device does go unresponsive, this'll create a rather annoying loop. I'd almost consider having a retry counter. If you write in a counter can you experiment a bit and see roughly how many retrys keeps your device active?

**Comment by marcelofg55** [comment]
Yes, you're right, it will write the data that was already on the buffer, Perhaps as a solution to this, I can change the PR to send the silent buffer in case of the device not responding, what do you think?

About the retry count you proposed, unless I'm missing something, it's already retrying as it is now. Since after the WaitForMultipleObjects timeouts it will call that again on the next loop.

**Comment by Andersama** [comment]
The code you're messing with right now is a thread which is responsible for ensuring audio data is written out to obs's internal system. I wrote this differently at some other point with only one wait call, but that wait is to ensure the device has written out a complete buffer of it's own.

Essentially, asio as a standard maintains a double buffer, I then decided to write a circular buffer to store that audio data temporarily, and then have a separate thread responsible for writing from that to obs's internal buffers. The two wait calls is to ensure at least one complete buffer from the device is available to write out*. It then loops and writes out all the data it can from the circular buffer.

It's not "retrying" it's in a permanent loop until the thread is killed when appropriate. Originally I thought a timeout event (which shouldn't happen w/ an infinite wait) would be a good sign to bail. Your change to the wait call means that a timeout event is par for the course, but it shouldn't happen that often. So the edit would be to have a retry counter, increment it on timeout events (reset to 0 on ordinary events) and then exit gracefully as before (if and only if the thread's timed out more than the retry limit in a row).

This is in part why I was confused as to why your edit fixed anything, the infinite wait shouldn't change anything. I also haven't touched obs's internals in ages so it might be doing something different now..

**Comment by marcelofg55** [comment]
Thanks for clarifying the retry thing, I understand your idea now, and yes I could add a retry count as an option so if anyone else is having this issue as well that setting could be useful.

By the way, if you want me to attach a video with some more details on the log, showing the before and after of the changes just let me know.

PD: This is my first PR here so I understand that you don't know me :). In case you want to check some previous work of mine I contributed to Godot's mostly with audio code, for example with a WASAP driver: https://github.com/godotengine/godot/blob/master/drivers/wasapi/audio_driver_wasapi.cpp

**Comment by Andersama** [comment]
No worries, was making sure I wasn't missing something in the video. If you make the edits for a retry counter and setting I'll be happy to merge it in.

Feel free to edit / push to the branch you're currently working on. It'll had the commits here so I can merge them in all at once.

**Comment by marcelofg55** [comment]
Hi, I've just pushed a new commit with the retry limit changes and I've also included a device timeout to make the device timeout handling completely optional. So this way the old behavior is not changed unless the user enables this new device timeout option.
Let me know if everything is okay with these new changes.

I've also been doing more tests and I found out that the timeout only occurs to me when using the ReaRoute ASIO driver. I have two audio interfaces and both of its drivers work fine, no timeout issues with them. So this new code can be helpful in case other users are also experiencing these timeouts with the ReaRoute driver (or maybe with other ASIO drivers as well).

File: src/asio-input.cpp [hunk 1]
```diff
@@ -608,15 +615,30 @@ class device_data {
 		int waitResult;
 
 		uint64_t buffer_time = ((device->frames * NSEC_PER_SEC) / device->samples_per_sec);
-		DWORD wait_time = DWORD((buffer_time / 1000000L) * 2);
+		DWORD wait_time = DWORD((buffer_time / 1000000L) * 8);
+		blog(LOG_INFO, "device %lu wait_time=%d", device->device_index, wait_time);
 		while (source && device) {
-			waitResult = WaitForMultipleObjects(3, signals_1, false, wait_time);
-			waitResult = WaitForMultipleObjects(3, signals_2, false, wait_time);
+			waitResult = WaitForMultipleObjects(3, signals_1, false, source->device_timeout ? wait_time : INFINITE);
```
Comment (Andersama) [inline src/asio-input.cpp:621]: Consider having the timeout value set once before these two calls, the two calls here are to allow a single buffer through, so we don't necessarily need two accesses and branches here.

File: src/asio-input.cpp [hunk 2]
```diff
@@ -1562,6 +1600,17 @@ obs_properties_t * asio_get_properties(void *unused)
 		"Warning: the real buffer returned by the device may differ";
 	obs_property_set_long_description(buffer_size, buffer_descr.c_str());
 
+	device_timeout = obs_properties_add_bool(props, "device timeout", obs_module_text("DeviceTimeout"));
+	std::string device_timeout_descr = "Device Timeout : Timeout if device takes too long to respond (might prevent 'Max audio buffering reached' problems).\n";
```
Comment (Andersama) [inline src/asio-input.cpp:1604]: Module text is grabbed from:
https://github.com/Andersama/obs-asio/blob/asiobass/src/data/locale/en-US.ini

If you're looking to add it and not have it fallback to it's default text you should add something in there as well, description can be the same.

Comment (Andersama) [reply src/asio-input.cpp:1604]: I'd shorten the name/description just to make it easier to display.
"Device May Timeout"
"Check this if your device has issues starting up / disconnecting. (May solve 'Max audio buffering reached' errors)."

**Comment by marcelofg55** [comment]
Hi, I've just changed what you've asked in your comments.
So let me know if I should do any other change and if what I've just changed is okay.

**Comment by Andersama** [comment]
Looks good 👍

**Pull request merged**